### ISSUE TITLE
wireguard-go data race

### ISSUE THREAD

**Issue by danderson** [tailscale/tailscale#225]
tl;dr: conn_linux.go needs a full rewrite. It's very racey and basically not possible to use safely. I could have fixed this one particular race (UpdateDst touching stuff without locking, and fmt.* printing the struct with reflection, also without locking), but that would be a band-aid for what needs to be a bigger surgery.

So, filing until I pop back up a few levels of yak shave.

```
Write at 0x00c00038e368 by goroutine 92:
116
  github.com/tailscale/wireguard-go/conn.(*NativeEndpoint).UpdateDst()
117
      /home/runner/work/corp/corp/wireguard-go/conn/conn_linux.go:103 +0x4dd
118
  github.com/tailscale/wireguard-go/device.(*Peer).SetEndpointAddress()
119
      /home/runner/work/corp/corp/wireguard-go/device/peer.go:335 +0x18f
120
  github.com/tailscale/wireguard-go/device.(*Device).RoutineHandshake()
121
      /home/runner/work/corp/corp/wireguard-go/device/receive.go:453 +0xea8
122

123
Previous read at 0x00c00038e368 by goroutine 101:
124
  fmt.(*pp).printValue()
125
      /opt/hostedtoolcache/go/1.13.7/x64/src/reflect/value.go:1914 +0x36a3
126
  fmt.(*pp).printValue()
127
      /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:869 +0xec7
128
  fmt.(*pp).printValue()
129
      /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:810 +0x283f
130
  fmt.(*pp).printValue()
131
      /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:880 +0x25da
132
  fmt.(*pp).printArg()
133
      /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:716 +0x2ee
134
  fmt.(*pp).doPrintf()
135
      /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:1030 +0x312
136
  fmt.Sprintf()
137
      /opt/hostedtoolcache/go/1.13.7/x64/src/fmt/print.go:219 +0x73
138
  log.(*Logger).Printf()
139
      /opt/hostedtoolcache/go/1.13.7/x64/src/log/log.go:179 +0x64
140
  github.com/tailscale/wireguard-go/device.(*Peer).SendHandshakeInitiation()
141
      /home/runner/work/corp/corp/wireguard-go/device/send.go:152 +0x2fe
```

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1171685499906294/1171686771219484) by [Unito](https://www.unito.io/learn-more)

**Comment by apenwarr** [comment]
Stupid question: are we actually using conn_linux.go? I thought we
were using the generic one + magicsock, but I can't 100% remember how
that worked.

**Comment by danderson** [comment]
Fixed in tailscale/wireguard-go@23aababa2084b7794e4b75ebbafe92e9d4b94404 .
